Tool packages

tcdent opened the pull request:
# Tool Packages

This PR introduces a new way of packaging tools in AgentStack. Tools are now wrapped in a framework-specific layer to allow for our implementations to remain completely generic. Some tools have been reimplemented to work with all frameworks instead of just importing from crewai-tools.

## Core Changes
- Tools now get wrapped in a framework-specific layer
- Reimplemented tools to work with all frameworks instead of just importing from crewai-tools
- Removed `tools_bundled` from ToolConfig as all tools are now bundled for namespacing
- `agentstack.tools` is not part of the public API and is to be used inside a user project to access tools

## Directory Search Implementation
As part of this effort, the directory search tool has been reimplemented as framework-agnostic functions:

- Added two core functions:
  - `search_directory(directory: str, query: str)`: Search any directory
  - `search_fixed_directory(query: str)`: Search preconfigured directory
- Removed dependency on CrewAI's BaseTool
- Uses embedchain.loaders.DirectoryLoader directly
- Framework-specific wrapping handled by frameworks.crewai.get_tool_callables

### Environment Variables
- `DIRECTORY_SEARCH_PATH`: Optional. Path to directory for fixed directory search.

## Composio Tool Implementation
The composio tool has been reimplemented as framework-agnostic functions:

- Added core functions:
  - `execute_action`: Execute specified Composio actions with parameters
  - `get_action_schema`: Retrieve schema for Composio actions
  - `find_actions_by_use_case`: Find actions based on use case
  - `find_actions_by_tags`: Locate actions by specified tags
- Removed dependency on CrewAI's BaseTool
- Uses composio library directly for core functionality
- Framework-specific wrapping handled automatically by agentstack.tools['composio']
- Updated config.json with proper dependencies and CTA field

### Environment Variables
- `COMPOSIO_API_KEY`: Required. API key for Composio service access.

## Stripe Tool Implementation
The Stripe tool has been reimplemented as framework-agnostic functions:
- Removed framework-specific dependencies
- Implemented core functions for:
  - Payment links (create/read)
  - Products (create/update)
  - Prices (create/update)
- Uses stripe library directly for core functionality
- Framework-specific wrapping handled automatically

### Environment Variables
- `STRIPE_SECRET_KEY`: Required. API key for Stripe service access.

Link to Devin run: https://app.devin.ai/sessions/24be273939424e7eacea38661326111e

tcdent committed:
Tool packaging abstractions. Needs tests.

tcdent committed:
Merge branch 'main' into tool-packages

tcdent committed:
Remove `packages` from ToolConfig tests.

tcdent committed:
Tool package init file placeholder.

tcdent committed:
Bug in ToolConfig.get_path

tcdent committed:
Merge cruft in tool_generation

tcdent committed:
Merge branch 'main' into tool-packages

tcdent committed:
Reformat tool packages to new format.

tcdent committed:
Merge branch 'main' into tool-packages

bboynton97 commented:
lets save this for 0.3 because it will require the project upgrade script. we can merge it now though and create a 0.2 branch for bug fixes and small features

tcdent committed:
Merge branch 'main' into tool-packages

tcdent committed:
Tools now get wrapped in a framework-specific layer to allow for our implementations to remain completely generic.
Reimplemented some tools to work with all frameworks instead of just importing from crewai-tools.
Removed `tools_bundled` from ToolCOnfig as all tools are now bundled for namespacing.
`agentstack.tools` is not part of the public API and is to be used inside a user project to access tools.

tcdent committed:
EnvFile now prefixes empty variables with a comment (#) to avoid overriding existing variables. More robust stripping on parse, too.

devin-ai-integration[bot] committed:
refactor: make directory_search tool framework-agnostic

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: make directory_search tool framework-agnostic

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
chore: add embedchain dependency to directory_search tool

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: make composio tool framework-agnostic

- Port implementation from crewAI-tools
- Make functions framework-agnostic
- Remove unnecessary framework-specific files
- Update config.json with correct dependencies

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
chore: restore cta field in composio tool config

- Add back cta field to indicate Composio's extensive tool offerings
- Update setup instructions to reference __init__.py

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: make vision tool framework-agnostic

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: namespace agent_connect environment variables with AGENT_CONNECT_ prefix

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: namespace agent_connect environment variables with AGENT_CONNECT_ prefix

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: resolve merge conflicts and namespace agent_connect environment variables

Co-Authored-By: root@a10k.co <root@a10k.co>

tcdent committed:
Prefix environment variables in agent_connect tool

devin-ai-integration[bot] committed:
refactor: make file_read tool framework-agnostic

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: move directory_search implementation to __init__.py

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: make stripe tool framework-agnostic

Co-Authored-By: root@a10k.co <root@a10k.co>

devin-ai-integration[bot] committed:
refactor: use stripe_agent_toolkit functions for stripe tool

Co-Authored-By: root@a10k.co <root@a10k.co>

tcdent committed:
Fix stripe config.json

tcdent committed:
Delete agentstack/tools/file_read/pyproject.toml

tcdent committed:
Reimplement stripe tool. Add __init__.py files to linter.

tcdent committed:
Removed .vscode/settings.json

tcdent committed:
Merge branch 'tcdent-gitignore' into tool-packages

tcdent committed:
Remove unused toml file.

tcdent committed:
Cleanup vision tool package

bboynton97 commented:
one challenge with this implementation is that by not copying the tool into the user's project, we remove the ability for them to modify the tool to best fit their use-case. So for Composio for example, there are a number of ways in which the tool can be adjusted for a specific implementation.

additionally, bundling all tools isn't the best idea. if a tool provider exports 20 tools and the developer only wants one of them, it consumes additional context to give the LLM all 20 tools, as well as creating the opportunity for the LLM to pick the wrong tool.

tcdent committed:
Prefix internal tools package to avoid confusion with the public `agentstack.tools` package

tcdent commented:
> users can still copy the tool source into their project tools dir to modify the implementation

We can add an `-e --editable` flag to copy the source to the user's project. 

I've been thinking about how we handle custom tools in templates. Still don't have a bright idea, but it would be cool to still be able to package up a project with custom tools and share the template. 

>> additionally, bundling all tools isn't the best idea. if a tool provider exports 20 tools and the developer only wants one of them, it consumes additional context to give the LLM all 20 tools, as well as creating the opportunity for the LLM to pick the wrong tool.

> will think about this. I wanted to incorporate namespacing so it was clear which tool the function was coming from. previously, installing a tool added all of the functions inside the tool package.

allowing the user to select which functions they're installing, in addition to the tool, using the new interactive `agentstack tools add` interface would be easy enough. 

`*agentstack.tools['tool_name']` for all tools
`agentstack.tools['tool_name']['function_name'], ...` for selected tools
`tools.function_name` for tools inside the user's project

just make the branching inside the framework ast editability a bit more complex, but very doable. 

moving the installed tools from `agentstack.json` to `conf/tools.yaml` would let us track this better

bboynton97 commented on agentstack/_tools/directory_search/config.json:
```diff
@@ -0,0 +1,12 @@
+  "tools": ["search_directory", "search_fixed_directory"],
+  "description": "Search through files in a directory using embedchain's DirectoryLoader",
+  "env": {
+    "DIRECTORY_SEARCH_TOOL_PATH": "Optional. Path to directory for fixed directory search."
```
`null`

bboynton97 commented on agentstack/_tools/firecrawl/config.json:
```diff
@@ -2,10 +2,12 @@
   "category": "browsing",
-  "packages": ["firecrawl-py"],
   "env": {
     "FIRECRAWL_API_KEY": "..."
```
`none`

bboynton97 commented on agentstack/_tools/neon/config.json:
```diff
@@ -2,10 +2,13 @@
     "url": "https://github.com/neondatabase/neon",
-    "packages": ["neon-api", "psycopg2-binary"],
     "env": {
         "NEON_API_KEY": "..."
```
`null`

bboynton97 commented on agentstack/_tools/~README.md:
```diff
@@ -1,5 +1,30 @@
-========================
+
+
+Tool Implementations
```
update to new structure

bboynton97 commented on agentstack/frameworks/crewai.py:
```diff
@@ -294,3 +324,20 @@ def remove_tool(tool: ToolConfig, agent_name: str):
+        assert tool_func.__doc__, f"Tool function {tool_func_name} is missing a docstring."
+
+        # apply the CrewAI tool decorator to the tool function
+        tool_funcs.append(_crewai_tool_decorator(tool_func))
```
🧼

bboynton97 commented on agentstack/generation/files.py:
```diff
@@ -1,4 +1,5 @@
 from typing import Optional, Union
+import string
```
merge this back

bboynton97 commented on agentstack/generation/tool_generation.py:
```diff
@@ -132,21 +54,11 @@ def remove_tool(tool_name: str, agents: Optional[list[str]] = []):
         print(term_color(f'Tool {tool_name} is not installed', 'red'))
         sys.exit(1)
 
+    # TODO ensure other agents are not using the tool
```
lets make an issue for this

bboynton97 commented on agentstack/generation/tool_generation.py:
```diff
@@ -132,21 +54,11 @@ def remove_tool(tool_name: str, agents: Optional[list[str]] = []):
-    except ValidationError as e:
-        print(term_color(f"Error removing tool:\n{e}", 'red'))
+    if tool.dependencies:
+        # TODO split on "==", ">=", etc. and only remove by package name
```
^

tcdent committed:
Merge branch 'main' into tool-packages

tcdent committed:
Support removing versioned dependencies when a tool is uninstalled.

tcdent committed:
Missing null values for placeholder tool env vars.

tcdent committed:
Fix malformed tool config.

tcdent committed:
Fix typing.

tcdent committed:
Add documentation for tool package structures.

braelyn-ai committed:
docs for adding tools

braelyn-ai committed:
Merge branch 'main' into tool-packages

# Conflicts:
#	agentstack/generation/tool_generation.py
#	agentstack/packaging.py
#	agentstack/tools.py

braelyn-ai reviewed (approved):
tl;dr
lgtm
